### PR TITLE
Make `HasStream` a super class of `HasWriter`

### PR DESCRIPTION
**Pull request by Ericson2314**
Does the good half of #63.

I went ahead to give that idea a spin, and found a few interesting things in the process. Will comment in line notes.

### CODE REVIEW

File: src/Capability/Stream.hs [hunk 1]
```diff
@@ -86,15 +86,34 @@ instance HasState tag [a] m => HasStream tag a (StreamStack m) where
 -- | Accumulate streamed values in forward order in a difference list.
 newtype StreamDList m (a :: *) = StreamDList (m a)
   deriving (Functor, Applicative, Monad, MonadIO, PrimMonad)
-instance HasWriter tag (DList a) m => HasStream tag a (StreamDList m) where
-  yield_ _ = coerce @(a -> m ()) $ tell @tag . DList.singleton
+instance HasStream tag (DList a) m => HasStream tag a (StreamDList m) where
```
Comment (Ericson2314) [inline src/Capability/Stream.hs]: There's no reason to pull in all of `Writer` to discharge this.

File: src/Capability/Stream.hs [hunk 2]
```diff
@@ -86,15 +86,34 @@ instance HasState tag [a] m => HasStream tag a (StreamStack m) where
...
+  yield_ _ = coerce @(a -> m ()) $ yield @tag . DList.singleton
   {-# INLINE yield_ #-}
 
+-- | Accumulate streamed values with their own monoid.
+newtype StreamLog m (a :: *) = StreamLog (m a)
+  deriving (Functor, Applicative, Monad, MonadIO, PrimMonad)
```
Comment (Ericson2314) [inline src/Capability/Stream.hs]: With this change, `WriterLog` needs an instance for `HasStream`, so the strategy becomes `StreamLog` and goes here. There is a `WriterLog` alias for back compat.

File: src/Capability/Stream.hs [hunk 3]
```diff
@@ -86,15 +86,34 @@ instance HasState tag [a] m => HasStream tag a (StreamStack m) where
...
+-- more efficient to apply 'Lift' to the underlying state capability. E.g.
+-- you should favour
+--
+-- > deriving (HasStream tag w) via
+-- >   StreamLog (Lift (SomeTrans (MonadState SomeStateMonad)))
+--
+-- over
+--
+-- > deriving (HasStream tag w) via
+-- >   Lift (SomeTrans (StreamLog (MonadState SomeStateMonad)))
```
Comment (Ericson2314) [inline src/Capability/Stream.hs]: This issue *already* existed: `Lift (StreamDList (WriterLog (... SomeStateMonad)))` was already possible, whereas `StreamDList (WriterLog (... (Lift SomeStateMonad)))` would be the preferred alternative. That said, with new `StreamLog`, less indirection is needed to create the same scenario. Given that, I decided to copy the warning from `Capability.Writer.Discouraged`, but not move the instance to a `Capability.Stream.Discouraged` as there's plenty of non-state-based ways to implement `HasStream`.

File: src/Capability/Writer.hs [hunk 4]
```diff
@@ -69,7 +68,7 @@ import GHC.Exts (Proxy#, proxy#)
 -- prop> listen @t (m >>= k) = listen @t m >>= \(a, w1) -> listen @t (k a) >>= \(b, w2) -> pure (b, w1 `mappend` w2)
 -- prop> pass @t (tell @t w >> pure (a, f)) = tell @t (f w) >> pure a
 -- prop> writer @t (a, w) = tell @t w >> pure a
-class (Monoid w, Monad m)
+class (Monoid w, Monad m, HasStream tag w m)
```
Comment (aherrmann) [inline src/Capability/Writer.hs]: I think `HasStream tag w m => HasWriter tag w m` deserves a good bit of additional documentation. To me this is far from obvious. (See overall review comment).

**Comment by aherrmann** [comment]
@Ericson2314 
> I'm quite willing to add all the docs you or anyone wants for these changes;

Thank you! Yes, I think this is a good addition to `capability`. I'll be happy to merge it with some more docs.

**Comment by Ericson2314** [comment]
OK added some documentation.

**Comment by aherrmann** [comment]
@Ericson2314 Sorry for leaving this hanging for so long. Thanks for the added documentation! Looks good to me. 

I noticed that the examples needed fixing. I added the corresponding changes.

**Comment by Ericson2314** [comment]
Oh, right. Thanks for fixing those!

**Pull request merged**